File: src/compression.cpp
#include <filesystem>
#include <set>
#include <fstream>
#include <algorithm>
#include <iostream>
#include "types.h"
#include "util.h"
#include "serialization.h"
namespace fs = std::filesystem;

/*
Common node format:
ID [unsigned int]
Node Type [bool]

Internal node format:
<Common node format>
Left node id [unsigned int]
Right node id [unsigned int]

Leaf node format:
<Common node format>
Character [char]
*/

Node* generate_tree(nodepq pq) {
    Node t;
    while (pq.size() > 1) {
        t = pq.top();
        Node *a = new Node;
        *a = t;
        pq.pop();
        t = pq.top();
        Node *b = new Node;
        *b = t;
        pq.pop();
        Node temp;
        temp.type = INTERNAL_NODE;
        temp.left = a;
        temp.right = b;
        
        temp.frequency = a->frequency + b->frequency;
        pq.push(temp);
    }
    t = pq.top();
    Node* root = new Node;
    *root = t;
    return root;
}

void gen_bitpair(Node& node, Node* parent, bitpair& bp, std::vector<unsigned char>& cur_bits, std::vector<unsigned char>& pseudo_bits) {
    switch (node.type) {
    case INTERNAL_NODE:
    {
        std::vector<unsigned char> a = cur_bits;
        a.push_back(0);
        gen_bitpair(*node.left, &node, bp, a, pseudo_bits);
        cur_bits.push_back(1);
        gen_bitpair(*node.right, &node, bp, cur_bits, pseudo_bits);
        break;
    }
    case PSEUDO_NODE:
        if (cur_bits.size() > 8) {
            parent->type = LEAF_NODE;
            if (parent->right == &node) {
                parent->chr = parent->left->chr;
                parent->frequency = parent->left->frequency;
            } else {
                parent->chr = parent->right->chr;
                parent->frequency = parent->right->frequency;
            }
            cur_bits.pop_back();
            bp[parent->chr] = cur_bits;
            delete parent->left;
            delete parent->right;
        }
        pseudo_bits = cur_bits;
        break;
    case LEAF_NODE:
        bp[node.chr] = cur_bits;
        break;
    }
}

void gen_bitpair(Node* node, bitpair& bp, std::vector<unsigned char>& pseudo_bits) {
    std::vector<unsigned char> temp;
    gen_bitpair(*node, nullptr, bp, temp, pseudo_bits);
}

void read_file(std::ifstream& file, long size, nodepq* pq) {
    std::map<char, unsigned int> charpair;
    file.seekg(0);
    char c;
    for (int i = 0; i < size; i++) {
        file.read(&c, sizeof(c));
        charpair[c]++;
    }
    for (const auto& elem : charpair)
        pq->push({ .type = LEAF_NODE, .frequency = elem.second, .chr = elem.first });
}

void compress_file(std::ofstream& file, const fs::path filename) {
    nodepq pq;
    write_string(filename.filename(), file);
    std::ifstream source(filename, std::ios::binary | std::ios::ate);
    unsigned long size = source.tellg();
    read_file(source, size, &pq);
    pq.push({ .type = PSEUDO_NODE, .frequency = 0 });
    Node* root = generate_tree(pq);
    bitpair bp;
    std::vector<unsigned char> pseudo_bits;
    gen_bitpair(root, bp, pseudo_bits);
    serialize_tree(root, file);
    serialize_text(bp, source, size, file, pseudo_bits);
    file.seekp(0, std::ios::end);
    free_tree(root);
}

void compress_file(const fs::path filename, std::string out_file) {
    std::ofstream file(out_file == "" ? filename.stem().replace_extension(".cmp").string() : out_file, std::ios::binary);
    char tmp = 0;
    file.write(&tmp, 1);
    compress_file(file, filename);
    file.close();
}

void compress_dir(std::ofstream& file, const fs::path path) {
    write_string(path.filename(), file);
    unsigned int offset = file.tellp();
    file.seekp(sizeof(unsigned int), std::ios::cur);
    unsigned int num_files = 0;
    for (const auto& entry : fs::directory_iterator(path)) {
        compress_file(file, entry.path());
        num_files++;
    }
    file.seekp(offset);
    file.write(reinterpret_cast<char*>(&num_files), sizeof(num_files));
}

void compress_dir(const fs::path path, std::string out_file) {
    std::ofstream file(out_file == "" ? path.stem().append(".cmp").string() : out_file, std::ios::binary);
    char tmp = 1;
    file.write(&tmp, 1);
    compress_dir(file, path);
    file.close();
}

File: src/compression.zig
const types = @import("types.zig");
const Node = types.Node;
const NodeType = types.NodeType;
const std = @import("std");
const fs = std.fs;
const log = std.log;
const Allocator = std.mem.Allocator;
const PriorityQueue = std.PriorityQueue(Node, void, lessThan);
const ArrayList = std.ArrayList;
const BitPair = std.AutoHashMap(u8, []u8);

pub fn compressDir(allocator: Allocator, dir_name: []const u8, out_file: *fs.File) !void {
    const writer = out_file.writer();
    try out_file.writeAll(dir_name);
    try writer.writeByte(0);
    const dir = try fs.openDirAbsolute(dir_name, .{});
    var iter = (try dir.openIterableDir(".", .{})).iterate();
    const num_files_offset = try out_file.getPos();
    try out_file.seekBy(@sizeOf(u32));
    var num_files: u32 = 0;
    while (try iter.next()) |e| : (num_files += 1) {
        log.info("Compressing {s}", .{e.name});
        var file = try dir.openFile(e.name, .{});
        defer file.close();
        try compressFile(allocator, e.name, file, out_file, writer);
    }
    try out_file.seekTo(num_files_offset);
    try writer.writeIntLittle(u32, num_files);
    log.info("Compressed {} files", .{num_files});
}

fn lessThan(context: void, a: Node, b: Node) std.math.Order {
    _ = context;
    return std.math.order(a.frequency, b.frequency);
}

fn compressFile(allocator: Allocator, filename: []const u8, file: fs.File, out_file: *fs.File, writer: fs.File.Writer) !void {
    _ = out_file;
    try writer.writeAll(filename);
    try writer.writeByte(0);
    var pq = PriorityQueue.init(allocator, {});
    defer pq.deinit();
    try pq.add(.{
        .node_type = NodeType.PSEUDO_NODE,
        .left = null,
        .right = null,
        .frequency = 0,
        .char = 0
    });
    try readFile(allocator, file, &pq);
    var root = try genTree(allocator, &pq);
    defer freeTree(allocator, root);
    var pseudo_bits: []u8 = undefined;
    var bitpair = try genBitPair(allocator, root, &pseudo_bits);
    defer allocator.free(pseudo_bits);
    defer {
        freeBitPair(allocator, bitpair);
        bitpair.deinit();
    }
    try serializeTree(root, writer);
    try file.seekTo(0);
    try serializeFile(allocator, bitpair, file, writer);
}

fn readFile(allocator: Allocator, file: fs.File, pq: *PriorityQueue) !void {
    var buffer = try allocator.alloc(u8, 4096);
    defer allocator.free(buffer);
    var l: usize = try file.readAll(buffer);
    var chars = std.AutoHashMap(u8, usize).init(allocator);
    defer chars.deinit();
    while (l > 0) : (l = try file.readAll(buffer)) {
        for (buffer[0..l]) |c| {
            const v = try chars.getOrPut(c);
            if (v.found_existing) {
                v.value_ptr.* += 1;
            } else {
                v.value_ptr.* = 1;
            }
        }
    }
    var iter = chars.keyIterator();
    while (iter.next()) |c| {
        try pq.add(.{
            .node_type = NodeType.LEAF_NODE,
            .frequency = chars.get(c.*).?,
            .left = null,
            .right = null,
            .char = c.*
        });
    }
}

fn genTree(allocator: Allocator, pq: *PriorityQueue) !*Node {
    while (pq.len > 1) {
        var a = try allocator.create(Node);
        a.* = pq.remove();
        var b = try allocator.create(Node);
        b.* = pq.remove();
        try pq.add(.{
            .node_type = NodeType.INTERNAL_NODE,
            .frequency = a.frequency + b.frequency,
            .left = a,
            .right = b,
            .char = 0
        });
    }
    const root = try allocator.create(Node);
    root.* = pq.remove();
    return root;
}

fn genBitPair(allocator: Allocator, root: *Node, pseudo_bits: *[]u8) !BitPair {
    var cur_bits = ArrayList(u8).init(allocator);
    defer cur_bits.deinit();
    var bitpair = BitPair.init(allocator);
    try genBitPairInner(allocator, root, null, &cur_bits, pseudo_bits, &bitpair);
    return bitpair;
}

fn genBitPairInner(allocator: Allocator, node: *Node, parent: ?*Node, cur_bits: *ArrayList(u8), pseudo_bits: *[]u8, bitpair: *BitPair) !void {
    switch (node.node_type) {
        NodeType.INTERNAL_NODE => {
            var a = try cur_bits.clone();
            defer a.deinit();
            try a.append(0);
            try genBitPairInner(allocator, node.left.?, node, &a, pseudo_bits, bitpair);
            var b = try cur_bits.clone();
            defer b.deinit();
            try b.append(1);
            try genBitPairInner(allocator, node.right.?, node, &b, pseudo_bits, bitpair);
        },
        NodeType.PSEUDO_NODE => {
            if (cur_bits.items.len > 8) {
                var parent_nn = parent.?;
                parent_nn.node_type = NodeType.LEAF_NODE;
                if (parent_nn.right.? == node) {
                    parent_nn.char = parent_nn.left.?.char;
                    parent_nn.frequency = parent_nn.left.?.frequency;
                } else {
                    parent_nn.char = parent_nn.right.?.char;
                    parent_nn.frequency = parent_nn.right.?.frequency;
                }
                _ = cur_bits.pop();
                try bitpair.put(parent_nn.char, cur_bits.items);
            }
            pseudo_bits.* = try allocator.alloc(u8, cur_bits.items.len);
            std.mem.copy(u8, pseudo_bits.*, cur_bits.items);
        },
        NodeType.LEAF_NODE => {
            var items: []u8 = try allocator.alloc(u8, cur_bits.items.len);
            std.mem.copy(u8, items, cur_bits.items);
            try bitpair.put(node.char, items);
        }
    }
}

fn freeTree(allocator: Allocator, root: *Node) void {
    if (root.node_type == .INTERNAL_NODE) {
        freeTree(allocator, root.left.?);
        freeTree(allocator, root.right.?);
    }
    allocator.destroy(root);
}

fn freeBitPair(allocator: Allocator, bitpair: BitPair) void {
    var iter = bitpair.valueIterator();
    while (iter.next()) |v| {
        allocator.free(v.*);
    }
}

fn serializeTree(root: *Node, writer: fs.File.Writer) !void {
    try writer.writeByte(@intFromEnum(root.node_type));
    switch (root.node_type) {
        .INTERNAL_NODE => {
            try serializeTree(root.left.?, writer);
            try serializeTree(root.right.?, writer);
        },
        .LEAF_NODE => try writer.writeByte(root.char),
        else => {}
    }
}

fn serializeFile(allocator: Allocator, bitpair: BitPair, in_file: fs.File, writer: fs.File.Writer) !void {
    var buffer = try allocator.alloc(u8, 4096);
    defer allocator.free(buffer);
    var byte: u8 = 0;
    var bit_idx: u4 = 0;
    var l: usize = try in_file.readAll(buffer);
    while (l > 0) : (l = try in_file.readAll(buffer)) {
        for (buffer[0..l]) |c| {
            for (bitpair.get(c).?) |b| {
                if (b == 1)
                    byte |= std.math.shl(u8, 1, bit_idx);
                bit_idx += 1;
                if (bit_idx == 8) {
                    try writer.writeByte(byte);
                    byte = 0;
                    bit_idx = 0;
                }
            }
        }
    }
}